detectable/src/main/java/com/synopsys/integration/detectable/detectables/go/gomod/GoModCommandRunner.java

```diff
@@ -0,0 +1,87 @@
+
+public class GoModCommandRunner {
+    // java:S5852: Warning about potential DoS risk.
+    @SuppressWarnings({ "java:S5852" })
+    private static final Pattern GENERATE_GO_LIST_JSON_OUTPUT_PATTERN = Pattern.compile("\\d+\\.[\\d.]+"); // Example: "go version go1.17.5 darwin/amd64" -> ""
+    private static final String JSON_OUTPUT_FLAG = "-json";
+    private static final String MODULE_OUTPUT_FLAG = "-m";
+
+    private final DetectableExecutableRunner executableRunner;
+
+    public GoModCommandRunner(DetectableExecutableRunner executableRunner) {
+        this.executableRunner = executableRunner;
+    }
+
+    // Excludes the "all" argument to return the root project modules
+    public List<String> runGoList(File directory, ExecutableTarget goExe) throws ExecutableFailedException {
+        return executableRunner.executeSuccessfully(ExecutableUtils.createFromTarget(directory, goExe, "list", MODULE_OUTPUT_FLAG, JSON_OUTPUT_FLAG))
+            .getStandardOutputAsList();
+    }
+
+    // TODO: Utilize the fields "Main": true, and "Indirect": true, fields from the JSON output to avoid running go list twice. Before switching to json output we needed to run twice. JM-01/2022
+    public List<String> runGoListAll(File directory, ExecutableTarget goExe) throws ExecutableFailedException {
+        boolean readOnlyFlagSupported;
+        try {
+            readOnlyFlagSupported = isReadOnlyFlagSupported(directory, goExe);
+        } catch (ExecutableFailedException e) {
+            // TODO: If we don't have a version for go, we don't do anything. This should probably result in a failure. JM-01/2022
+            return new ArrayList<>();
+        }
+
+        List<String> goListCommand = new LinkedList<>();
+        goListCommand.add("list");
+        if (readOnlyFlagSupported) {
+            // Providing a readonly flag prevents the command from modifying customer's source.
+            goListCommand.add("-mod=readonly");
+        }
+        goListCommand.addAll(Arrays.asList(MODULE_OUTPUT_FLAG, JSON_OUTPUT_FLAG, "all"));
+
+        return executableRunner.executeSuccessfully(ExecutableUtils.createFromTarget(directory, goExe, goListCommand))
+            .getStandardOutputAsList();
+    }
+
+    private boolean isReadOnlyFlagSupported(File directory, ExecutableTarget goExe) throws ExecutableFailedException {
+        List<String> goVersionOutput = executableRunner.executeSuccessfully(ExecutableUtils.createFromTarget(directory, goExe, "version"))
+            .getStandardOutputAsList();
+        Matcher matcher = GENERATE_GO_LIST_JSON_OUTPUT_PATTERN.matcher(goVersionOutput.get(0));
+        if (matcher.find()) {
+            String version = matcher.group(); // 1.16.5
+            String[] parts = version.split("\\.");
+            boolean majorVersionBeyond1 = Integer.parseInt(parts[0]) > 1;
+            boolean minorVersion14AndUp = Integer.parseInt(parts[1]) >= 14;
+            return majorVersionBeyond1 || minorVersion14AndUp;
+        }
```

Comment:
I had to step through the original logic in a debugger to understand what's going on. 
This is a bit better, but no great. Do we have a SerVer library we could use?

Reply:
First I'd recomment splitting the parsing from the logic. Parse it to a GoModVersion(major, minor) and then in here do GoModVersion.Major > 1. I'm not aware of a semver parser that we have. Additionally, I'd have nice tests on the parser. Its not exactly clear if you have to support many different version string schemes, but the given example could be parsed much simpler, say 'if startsWith("go version go") then substring(10)' and then removeToString(" ") and then split(".") seems like it would get you there.